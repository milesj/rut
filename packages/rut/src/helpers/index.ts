export * from './predicates';
export * from './utils';
