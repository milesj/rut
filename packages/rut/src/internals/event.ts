/* eslint-disable complexity, max-classes-per-file */

export class BaseEvent {
  bubbles: boolean = true;

  cancelable: boolean = true;

  currentTarget: unknown = {};

  defaultPrevented: boolean = false;

  eventPhase: number = 0;

  isTrusted: boolean = true;

  propagationStopped: boolean = false;

  srcElement: null = null;

  target: unknown = {};

  timeStamp: number;

  type: string;

  constructor(type: string) {
    this.timeStamp = Date.now();
    this.type = type;
  }

  preventDefault(): void {
    this.defaultPrevented = true;
  }

  stopImmediatePropagation(): void {
    this.propagationStopped = true;
  }

  stopPropagation(): void {
    this.propagationStopped = true;
  }
}

export class SyntheticEvent extends BaseEvent {
  nativeEvent: Event;

  persisted: boolean = false;

  constructor(type: string, nativeEvent: Event) {
    super(type);

    this.nativeEvent = nativeEvent;

    // Copy over non-function properties
    Object.entries(nativeEvent).forEach(([key, value]) => {
      if (typeof value !== 'function') {
        Object.defineProperty(this, key, {
          enumerable: true,
          value,
        });
      }
    });
  }

  isDefaultPrevented(): boolean {
    return this.defaultPrevented;
  }

  isPersistent(): boolean {
    return this.persisted;
  }

  isPropagationStopped(): boolean {
    return this.propagationStopped;
  }

  persist(): void {
    this.persisted = true;
  }

  preventDefault(): void {
    this.nativeEvent.preventDefault();
    this.defaultPrevented = true;
  }

  stopPropagation(): void {
    this.nativeEvent.stopPropagation();
    this.propagationStopped = true;
  }
}
