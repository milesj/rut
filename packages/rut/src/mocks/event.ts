/* eslint-disable complexity, max-classes-per-file */

/**
 * In both the `mockEvent` and `mockSyntheticEvent` functions below,
 * you'll notice that every return is being type ignored.
 * We are doing this so that the generic type takes precendence,
 * as we want inferrence to be predominantly used.
 *
 * For example, if a test is emitting `findOne('button').emit('click`)`,
 * the first argument is typed as `React.MouseEvent<HTMLButtonElement, MouseEvent>`.
 * We can take advantage of type inferrence by mocking the argument at
 * the call site, like so: `findOne('button').emit('click', {}, mockSyntheticEvent('click'))`.
 *
 * With this pattern, mocks are easily typed, and the underlying event object
 * structure is close enough for most, if not all of test cases.
 */

interface EventOptions {
  currentTarget?: HTMLElement;
  target?: Element;
}

export class BaseEvent {
  bubbles: boolean = true;

  cancelable: boolean = true;

  currentTarget: unknown = {};

  defaultPrevented: boolean = false;

  eventPhase: number = 0;

  isTrusted: boolean = true;

  propagationStopped: boolean = false;

  srcElement: null = null;

  target: unknown = {};

  timeStamp: number;

  type: string;

  constructor(type: string) {
    this.timeStamp = Date.now();
    this.type = type;
  }

  preventDefault(): void {
    this.defaultPrevented = true;
  }

  stopImmediatePropagation(): void {
    this.propagationStopped = true;
  }

  stopPropagation(): void {
    this.propagationStopped = true;
  }
}

// https://developer.mozilla.org/en-US/docs/Web/Events
// istanbul ignore next
function createHostEvent(type: string): Event {
  switch (type) {
    case 'animationcancel':
    case 'animationiteration':
    case 'animationend':
    case 'animationstart':
      return new AnimationEvent(type);
    case 'beforeunload':
      return new BeforeUnloadEvent();
    case 'auxclick':
    case 'click':
    case 'contextmenu':
    case 'dblclick':
    case 'mousedown':
    case 'mouseenter':
    case 'mouseleave':
    case 'mousemove':
    case 'mouseout':
    case 'mouseover':
    case 'mouseup':
      return new MouseEvent(type);
    case 'devicelight':
    case 'devicemotion':
    case 'deviceorientation':
    case 'deviceorientationabsolute':
      return new DeviceLightEvent(type);
    case 'blur':
    case 'focus':
      return new FocusEvent(type);
    case 'copy':
    case 'cut':
    case 'paste':
      return new ClipboardEvent(type);
    case 'compositionend':
    case 'compositionstart':
    case 'compositionupdate':
      return new CompositionEvent(type);
    case 'drag':
    case 'dragend':
    case 'dragenter':
    case 'dragexit':
    case 'dragleave':
    case 'dragover':
    case 'dragstart':
    case 'drop':
      return new DragEvent(type);
    case 'error':
      return new ErrorEvent(type);
    case 'hashchange':
      return new HashChangeEvent(type);
    case 'keydown':
    case 'keypress':
    case 'keyup':
      return new KeyboardEvent(type);
    case 'loadend':
    case 'progress':
    case 'readystatechange':
      return new ProgressEvent(type);
    case 'message':
    case 'messageerror':
      return new MessageEvent(type);
    case 'pagehide':
    case 'pageshow':
      return new PageTransitionEvent();
    case 'gotpointercapture':
    case 'lostpointercapture':
    case 'pointercancel':
    case 'pointerdown':
    case 'pointerenter':
    case 'pointerleave':
    case 'pointerover':
    case 'pointermove':
    case 'pointerout':
    case 'pointerup':
      return new PointerEvent(type);
    case 'popstate':
      return new PopStateEvent(type);
    case 'touchcancel':
    case 'touchend':
    case 'touchmove':
    case 'touchstart':
      return new TouchEvent(type);
    case 'transitioncancel':
    case 'transitionend':
    case 'transitionrun':
    case 'transitionstart':
      return new TransitionEvent(type);
    case 'abort':
    case 'scroll':
    case 'resize':
      return new UIEvent(type);
    case 'securitypolicyviolation':
      return new SecurityPolicyViolationEvent(type);
    case 'storage':
      return new StorageEvent(type);
    case 'unhandledrejection':
      return new PromiseRejectionEvent(type, { promise: Promise.resolve() });
    case 'wheel':
      return new WheelEvent(type);
    default:
      return new Event(type);
  }
}

export function mockEvent<T = Event>(type: string, options: EventOptions = {}): T {
  let event: Event;

  // JSDOM environment does not exist, which means we do not have events.
  // Return a very custom low-level event object for the time being.
  if (typeof window === 'undefined') {
    // @ts-ignore Ignore legacy fields
    event = new BaseEvent(type);
  } else {
    event = createHostEvent(type);
  }

  if (options.target) {
    Object.defineProperty(event, 'target', {
      value: options.target,
    });
  }

  if (options.currentTarget || options.target) {
    Object.defineProperty(event, 'currentTarget', {
      value: options.currentTarget || options.target,
    });
  }

  // @ts-ignore
  return event;
}

export class SyntheticEvent extends BaseEvent {
  nativeEvent: Event;

  persisted: boolean = false;

  constructor(type: string, nativeEvent: Event) {
    super(type);

    this.nativeEvent = nativeEvent;
    this.bubbles = nativeEvent.bubbles;
    this.cancelable = nativeEvent.cancelable;
    this.currentTarget = nativeEvent.currentTarget;
    this.defaultPrevented = nativeEvent.defaultPrevented;
    this.eventPhase = nativeEvent.eventPhase;
    this.isTrusted = nativeEvent.isTrusted;
    this.target = nativeEvent.target;
    this.timeStamp = nativeEvent.timeStamp;
  }

  isDefaultPrevented(): boolean {
    return this.defaultPrevented;
  }

  isPersistent(): boolean {
    return this.persisted;
  }

  isPropagationStopped(): boolean {
    return this.propagationStopped;
  }

  persist(): void {
    this.persisted = true;
  }

  preventDefault(): void {
    this.nativeEvent.preventDefault();
    this.defaultPrevented = true;
  }

  stopPropagation(): void {
    this.nativeEvent.stopPropagation();
    this.propagationStopped = true;
  }
}

export function mockSyntheticEvent<T = React.SyntheticEvent>(
  type: string,
  options: EventOptions = {},
): T {
  // @ts-ignore
  return new SyntheticEvent(type, mockEvent(type, options));
}
