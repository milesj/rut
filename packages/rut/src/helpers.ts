/* eslint-disable complexity */

import React from 'react';
import * as ReactIs from 'react-is';
import { UnknownProps } from './types';
import { doAsyncAct } from './internals/act';

export interface NodeLike {
  $$typeof: symbol | number;
  type?: NodeLike;
  props?: UnknownProps;
}

export interface ContextLike extends NodeLike {
  _context: {
    displayName?: string;
  };
}

export interface PortalLike extends NodeLike {
  containerInfo: HTMLElement;
}

/**
 * Return true if the value is either a React class or component declaration.
 */
export function isReactComponent(value: unknown): value is React.ComponentType {
  return typeof value === 'function';
}

/**
 * Return true if the value is a React class component instance.
 */
export function isReactComponentInstance(value: unknown): value is Function {
  return (
    typeof value === 'object' && !!value && 'isReactComponent' in value && 'constructor' in value
  );
}

/**
 * Return true if the value is structurally similar to a React element or node,
 * by checking for the existance of `$$typeof`.
 */
export function isReactNodeLike(value: unknown): value is NodeLike {
  return typeof value === 'object' && !!value && '$$typeof' in value;
}

/**
 * Return the `displayName` of a React context node, otherwise return "Context".
 */
export function getContextName(node: NodeLike): string {
  if (node.type) {
    return getContextName(node.type);
  }

  return ('_context' in node && (node as ContextLike)._context.displayName) || 'Context';
}

/**
 * Return the name of a React component, element, or node, by inferring metadata from
 * `$$typeof`. Utilizes the `react-is` package for common scenarios, but also takes
 * into account fibers and nodes not wrapped with React elements.
 */
export function getTypeName(type: unknown): string {
  if (!type) {
    return 'UNKNOWN';
  }

  if (isReactComponent(type)) {
    return type.displayName || type.name;
  }

  if (isReactComponentInstance(type)) {
    return getTypeName(type.constructor);
  }

  if (!isReactNodeLike(type) || typeof type === 'string') {
    return String(type);
  }

  const typeOf = type.$$typeof;

  if (ReactIs.isContextConsumer(type) || typeOf === ReactIs.ContextConsumer) {
    return `${getContextName(type)}.Consumer`;
  }

  if (ReactIs.isContextProvider(type) || typeOf === ReactIs.ContextProvider) {
    return `${getContextName(type)}.Provider`;
  }

  if (ReactIs.isForwardRef(type) || typeOf === ReactIs.ForwardRef) {
    return 'ForwardRef'; // We lose the component name
  }

  if (ReactIs.isFragment(type) || typeOf === ReactIs.Fragment) {
    return 'Fragment';
  }

  if (ReactIs.isLazy(type) || typeOf === ReactIs.Lazy) {
    return 'Lazy';
  }

  if (ReactIs.isMemo(type) || typeOf === ReactIs.Memo) {
    return `Memo(${getTypeName(type.type)})`;
  }

  if (ReactIs.isProfiler(type) || typeOf === ReactIs.Profiler) {
    return `Profiler(${type.props!.id})`;
  }

  if (ReactIs.isPortal(type) || typeOf === ReactIs.Portal) {
    const portal = type as PortalLike;

    return `Portal(${portal.containerInfo.id || portal.containerInfo.tagName.toLowerCase()})`;
  }

  if (ReactIs.isStrictMode(type) || typeOf === ReactIs.StrictMode) {
    return 'StrictMode';
  }

  if (ReactIs.isSuspense(type) || typeOf === ReactIs.Suspense) {
    return 'Suspense';
  }

  if (ReactIs.isElement(type)) {
    return getTypeName(type.type);
  }

  return String(type);
}

/**
 * Format a value for use within error messages.
 */
export function formatValue(value: unknown): string {
  const typeOf = typeof value;

  if (typeOf === 'string') {
    return `"${value}"`;
  } else if (typeOf === 'number' || typeOf === 'boolean' || value === null || value === undefined) {
    return `\`${String(value)}\``;
  }

  return `\`${getTypeName(value)}\``;
}

/**
 * Wait for async calls to finish by delaying execution.
 */
export async function wait(delay: number = 10): Promise<void> {
  await doAsyncAct(async () => {
    await new Promise(resolve => {
      setTimeout(resolve, delay);
    });
  });
}
