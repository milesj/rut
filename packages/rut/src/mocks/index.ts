import { mockEvent, mockSyntheticEvent } from './event';

export { mockEvent, mockSyntheticEvent };
