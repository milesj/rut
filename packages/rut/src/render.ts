import React from 'react';
import AsyncResult from './AsyncResult';
import SyncResult from './SyncResult';
import { globalOptions } from './configure';
import { doAct, doAsyncAct } from './internals/act';
import { RendererOptions } from './types';

// export function render<Props extends object = {}>(
//   element: React.ReactElement<Props>,
//   options?: RendererOptions,
// ): SyncResult<Props> {
//   const opts = {
//     ...globalOptions,
//     ...options,
//   };

//   return doAct(() => new SyncResult(element, opts));
// }

export async function renderAndWait<Props extends object = {}>(
  element: React.ReactElement<Props>,
  options?: RendererOptions,
): Promise<AsyncResult<Props>> {
  const opts = {
    ...globalOptions,
    ...options,
  };

  return doAsyncAct(() => new AsyncResult(element, opts), opts.asyncMode);
}

export const render = renderAndWait;
