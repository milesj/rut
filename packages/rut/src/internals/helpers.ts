import React from 'react';
import Element from '../Element';
import { isReactNodeLike, NodeLike } from '../helpers';
import { UnknownProps } from '../types';

export function checkIsRutElement(value: unknown) {
  if (value instanceof Element) {
    return;
  }

  if (
    typeof value === 'object' &&
    value !== null &&
    (value.constructor.name === 'Element' || (value as Element).isRutElement === true)
  ) {
    return;
  }

  throw new Error('Expected a Rut `Element`.');
}

export function getPropFromElement<P, K extends keyof P>(
  element: Element<P>,
  name: K,
): P[K] | undefined {
  // @ts-ignore Allow internal access
  return element.element.props[name];
}

export function getPropForEmitting<P, K extends keyof P>(element: Element<P>, name: K): P[K] {
  let eventName = String(name);

  if (!eventName.startsWith('on')) {
    eventName = `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
  }

  const prop = getPropFromElement(element, eventName as K);

  if (!prop) {
    throw new Error(`Prop \`${eventName}\` does not exist.`);
  } else if (typeof prop !== 'function') {
    throw new TypeError(`Prop \`${eventName}\` is not a function.`);
    // @ts-ignore Allow internal access
  } else if (typeof element.element.type !== 'string') {
    throw new TypeError('Emitting events is only allowed on host components (DOM elements).');
  }

  return prop;
}

// Keep shallow equal in sync with React core!
// https://github.com/facebook/react/blob/master/packages/shared/shallowEqual.js
export function shallowEqual(objA: unknown, objB: unknown): boolean {
  if (Object.is(objA, objB)) {
    return true;
  }

  if (typeof objA !== 'object' || objA === null || typeof objB !== 'object' || objB === null) {
    return false;
  }

  const keysA = Object.keys(objA);
  const keysB = Object.keys(objB);

  if (keysA.length !== keysB.length) {
    return false;
  }

  // Test for A's keys different from B.
  // eslint-disable-next-line unicorn/no-for-loop
  for (let i = 0; i < keysA.length; i += 1) {
    if (
      !Object.prototype.hasOwnProperty.call(objB, keysA[i]) ||
      // @ts-ignore
      !Object.is(objA[keysA[i]], objB[keysA[i]])
    ) {
      return false;
    }
  }

  return true;
}

export function containsProps(props: UnknownProps, contains: UnknownProps): boolean {
  return Object.keys(contains).every(prop => props[prop] === contains[prop]);
}

/**
 * Memo and other exotics do not appear in the reconciled tree,
 * so we need to remove it and dig deeper to find the true element type.
 */
export function unwrapExoticType(element: NodeLike): React.ReactElement['type'] {
  const rootType = element.type!;

  if (typeof rootType === 'string' || typeof rootType === 'function') {
    return rootType;
  }

  if (isReactNodeLike(rootType)) {
    return unwrapExoticType(rootType);
  }

  return rootType;
}
