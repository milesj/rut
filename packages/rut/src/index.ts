/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { act } from 'react-test-renderer';
import matchers from './matchers';
import { render, renderAndWait } from './render';
import mockSyntheticEvent from './mocks/event';

export * from './helpers';
export * from './types';

export { act, matchers, mockSyntheticEvent, render, renderAndWait };
