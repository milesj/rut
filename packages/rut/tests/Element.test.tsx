import React, { useState } from 'react';
import { render } from '../src/render';
import { mockSyntheticEvent } from '../src/mocks/event';
import {
  FuncComp,
  FuncCompWithDisplayName,
  ClassComp,
  ClassCompWithDisplayName,
  TestProps,
} from './fixtures';
import { runAsyncCall } from './helpers';

describe('Element', () => {
  it('can serialize as a snapshot', () => {
    const { root } = render<TestProps>(
      <FuncComp>
        <b>Child</b>
      </FuncComp>,
    );

    expect(root).toMatchSnapshot();
  });

  describe('debug()', () => {
    it('debugs based on element depth', () => {
      const { root } = render<{}>(
        <div>
          <section>
            <article>
              <h1>Title</h1>
            </article>
          </section>
        </div>,
      );

      expect(root.debug({ log: false })).toMatchSnapshot();
      expect(root.findOne('section').debug({ log: false })).toMatchSnapshot();
      expect(root.findOne('h1').debug({ log: false })).toMatchSnapshot();
    });
  });

  describe('dispatch()', () => {
    it('errors if prop does not exist', () => {
      expect(() => {
        const { root } = render(
          <div>
            <span />
          </div>,
        );

        // @ts-ignore
        root.findOne('span').dispatch('onFake');
      }).toThrowError('Prop `onFake` does not exist.');
    });

    it('errors if prop is not a function', () => {
      expect(() => {
        const { root } = render(
          <div>
            <span id="foo" />
          </div>,
        );

        // @ts-ignore
        root.findOne('span').dispatch('id');
      }).toThrowError('Prop `id` is not a function.');
    });

    it('errors if dispatching on a non-host component', () => {
      interface DispatchProps {
        onSomething: () => void;
      }

      function DispatchComp(props: DispatchProps) {
        return <div />;
      }

      expect(() => {
        const { root } = render<DispatchProps>(<DispatchComp onSomething={() => {}} />);

        // @ts-ignore
        root.dispatch('onSomething');
      }).toThrowError('Dispatching events is only allowed on host components (DOM elements).');
    });

    it('executes the function prop', () => {
      const spy = jest.fn();

      function DispatchComp() {
        return (
          <button type="button" onClick={spy}>
            Click
          </button>
        );
      }

      const { root } = render<{}>(<DispatchComp />);

      root.findOne('button').dispatch('onClick');

      expect(spy).toHaveBeenCalledWith(expect.any(Object));
    });
  });

  describe('dispatchAndWait()', () => {
    function DispatchTest() {
      const [count, setCount] = useState(0);
      const onClick = () =>
        runAsyncCall(() => {
          setCount(count + 1);
        });

      return (
        <div>
          <span>{count}</span>
          <button type="button" onClick={onClick}>
            Increment
          </button>
        </div>
      );
    }

    it('waits for the async and re-render', async () => {
      const { root } = render<{}>(<DispatchTest />);

      expect(root).toContainNode(0);

      await root.findOne('button').dispatchAndWait('onClick', mockSyntheticEvent('onClick'));

      expect(root).toContainNode(1);
    });
  });

  describe('find()', () => {
    it('returns an empty array if none found', () => {
      const { root } = render(<div />);

      expect(root.find('span')).toEqual([]);
    });

    it('returns all HTML elements by name', () => {
      const { root } = render(
        <div>
          <span>1</span>
          <span>2</span>
          <span>3</span>
        </div>,
      );

      expect(root.find('span')).toHaveLength(3);
    });

    it('returns all HTML elements at any depth', () => {
      const { root } = render(
        <div>
          <span>
            1
            <span>
              2<span>3</span>
            </span>
          </span>
        </div>,
      );

      expect(root.find('span')).toHaveLength(3);
    });

    it('returns all components by type', () => {
      const { root } = render(
        <div>
          <FuncComp>1</FuncComp>
          <FuncComp>2</FuncComp>
          <FuncComp>3</FuncComp>
        </div>,
      );

      expect(root.find(FuncComp)).toHaveLength(3);
    });

    it('returns all components by type at any depth', () => {
      const { root } = render(
        <div>
          <FuncComp>
            1
            <FuncComp>
              2<FuncComp>3</FuncComp>
            </FuncComp>
          </FuncComp>
        </div>,
      );

      expect(root.find(FuncComp)).toHaveLength(3);
    });

    it('filters found elements based on defined props', () => {
      const { root } = render(
        <form>
          <input type="text" name="name" />
          <input type="email" name="email" />
          <input type="password" name="password" />
        </form>,
      );

      expect(root.find('input', { name: 'email' })).toHaveLength(1);
    });
  });

  describe('findAt()', () => {
    it('returns the element at each index', () => {
      const { root } = render(
        <div>
          <span>1</span>
          <span>2</span>
          <span>3</span>
        </div>,
      );

      const one = root.findAt('span', 'first');
      const two = root.findAt('span', 1);
      const three = root.findAt('span', 'last');

      expect(one).toContainNode('1');
      expect(two).toContainNode('2');
      expect(three).toContainNode('3');
    });

    it('errors if unknown type passed', () => {
      const { root } = render(<div />);

      expect(() => {
        // @ts-ignore Allow invalid
        root.findAt('span', 'middle');
      }).toThrowError('Invalid index type "middle".');
    });

    it('errors if first not found', () => {
      const { root } = render(<div />);

      expect(() => {
        root.findAt('span', 'first');
      }).toThrowError('Expected to find an element at index 0 for `span`.');
    });

    it('errors if last not found', () => {
      const { root } = render(<div />);

      expect(() => {
        root.findAt('span', 'last');
      }).toThrowError('Expected to find an element at index -1 for `span`.');
    });

    it('errors if index not found', () => {
      const { root } = render(<div />);

      expect(() => {
        root.findAt('span', 3);
      }).toThrowError('Expected to find an element at index 3 for `span`.');
    });
  });

  describe('findOne()', () => {
    it('errors if no result found', () => {
      const { root } = render(<div />);

      expect(() => {
        root.findOne('span');
      }).toThrowError('Expected to find 1 element for `span`, found 0.');
    });

    it('errors if too many results found', () => {
      const { root } = render(
        <div>
          <span>1</span>
          <span>2</span>
          <span>3</span>
        </div>,
      );

      expect(() => {
        root.findOne('span');
      }).toThrowError('Expected to find 1 element for `span`, found 3.');
    });

    it('returns the HTML element found by name', () => {
      const { root } = render(
        <div>
          <span>1</span>
        </div>,
      );

      expect(root.findOne('span')).toContainNode(1);
    });

    it('returns the first component by type found by name', () => {
      const { root } = render(
        <div>
          <FuncComp>1</FuncComp>
        </div>,
      );

      expect(root.findOne(FuncComp)).toContainNode(1);
    });
  });

  describe('name()', () => {
    it('returns HTML tag', () => {
      const { root } = render(<div />);

      expect(root.name()).toBe('div');
    });

    it('returns function component name', () => {
      const { root } = render<TestProps>(<FuncComp />);

      expect(root.name()).toBe('FuncComp');
    });

    it('returns function component display name', () => {
      const { root } = render<TestProps>(<FuncCompWithDisplayName />);

      expect(root.name()).toBe('CustomFuncName');
    });

    it('returns class component name', () => {
      const { root } = render<TestProps>(<ClassComp />);

      expect(root.name()).toBe('ClassComp');
    });

    it('returns class component display name', () => {
      const { root } = render<TestProps>(<ClassCompWithDisplayName />);

      expect(root.name()).toBe('CustomCompName');
    });

    it('returns component name with HOC', () => {
      function connect<P>(WrappedComponent: React.ComponentType<P>) {
        function Connect(props: P) {
          return <WrappedComponent {...props} />;
        }

        Connect.displayName = `connect(${WrappedComponent.displayName || WrappedComponent.name})`;

        return Connect;
      }

      const Connected = connect(ClassComp);

      const { root } = render<{}>(<Connected />);

      expect(root.name()).toBe('connect(ClassComp)');
    });
  });
});
