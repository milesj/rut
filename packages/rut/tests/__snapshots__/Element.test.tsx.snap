// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element can serialize as a snapshot 1`] = `
<FuncComp>
  <span>
    <b>Child</b>
  </span>
</FuncComp>
`;

exports[`Element debug() debugs based on element depth 1`] = `
"<div>
  <section>
    <article>
      <h1>Title</h1>
    </article>
  </section>
</div>"
`;

exports[`Element debug() debugs based on element depth 2`] = `
"<section>
  <article>
    <h1>Title</h1>
  </article>
</section>"
`;

exports[`Element debug() debugs based on element depth 3`] = `"<h1>Title</h1>"`;
